### PR TITLE
Implement core module basics [#9] - helpers implementation

### PR DESCRIPTION
**Pull request by rebzden**

### CODE REVIEW

File: packages/core/src/core.ts [hunk 1]
```diff
@@ -1 +1,49 @@
-export const x = true
+export const isEmpty = value => {
```
Comment (ulfryk) [inline packages/core/src/core.ts]: what type is `value` ?

File: packages/core/src/core.ts [hunk 2]
```diff
@@ -1 +1,49 @@
...
+  if (isNothing(value) || value === "") {
```
Comment (ulfryk) [inline packages/core/src/core.ts]: let's use `'` always

File: packages/core/src/core.ts [hunk 3]
```diff
@@ -1 +1,49 @@
...
+    return Object.keys(value).length === 0;
+  }
+  return false;
+};
+
+export const noop = () => {};
+
+export const isNothing = value => {
+  return value == null;
+};
```
Comment (ulfryk) [inline packages/core/src/core.ts]: ` => { return x } ` --> ` => x `

File: packages/core/src/core.ts [hunk 4]
```diff
@@ -1 +1,49 @@
...
+  }
+  if (typeof value === "object") {
+    return Object.keys(value).length === 0;
+  }
+  return false;
+};
+
+export const noop = () => {};
+
+export const isNothing = value => {
```
Comment (ulfryk) [inline packages/core/src/core.ts]: arg type

File: packages/core/src/core.ts [hunk 5]
```diff
@@ -1 +1,49 @@
...
+export const areEqual = (a: any, b: any) => {
+  if (a === b || (a !== a && b !== b)) {
+    return true;
+  }
+
+  if (!a || !b) {
+    return false;
+  }
+
+  if (isFunction(a.equals) && isFunction(b.equals)) {
```
Comment (ulfryk) [inline packages/core/src/core.ts]: i'd suggest adding interface defining [Setoid](https://github.com/fantasyland/fantasy-land#setoid) and function `isSetoid(a: ?) a is Setoid` to use here ;)

File: packages/core/src/core.ts [hunk 6]
```diff
@@ -1 +1,49 @@
...
+  }
+  return false;
+};
+
+export const isFunction = (f: any) => {
+  return Boolean(f && f.constructor && f.call && f.apply);
+};
+
+export const compose = (f: Function, g: Function) => {
+  return function(x: unknown) {
```
Comment (ulfryk) [inline packages/core/src/core.ts]: `(f: ?, g: ?) => (x: ?): ? => f(g(x))`

- remove obsolete `function` and `return` statements
- add proper generic types here ;)

File: packages/core/src/core.ts [hunk 7]
```diff
@@ -1 +1,49 @@
...
+    return f(g(x));
+  };
+};
```
Comment (ulfryk) [inline packages/core/src/core.ts]: - split this file please - each helper in it's own file
- Setoid and it's related functions grouped into one file

**Review by ulfryk — CHANGES_REQUESTED** [review]

**Comment by ulfryk** [comment]
Maybe add issue number to commit/PR title (it helps to tie things together later)

File: packages/core/src/core.ts (hunk 1 above)
Comment (rebzden) [reply packages/core/src/core.ts]: fixed

File: packages/core/src/core.ts (hunk 2 above)
Comment (rebzden) [reply packages/core/src/core.ts]: fixed

File: packages/core/src/core.ts (hunk 4 above)
Comment (rebzden) [reply packages/core/src/core.ts]: fixed

File: packages/core/src/core.ts (hunk 5 above)
Comment (rebzden) [reply packages/core/src/core.ts]: will do in the future

File: packages/core/src/core.ts (hunk 6 above)
Comment (rebzden) [reply packages/core/src/core.ts]: split the file, generic will be added with more infrastructure added later

File: packages/core/src/core.ts (hunk 7 above)
Comment (rebzden) [reply packages/core/src/core.ts]: as above

File: packages/core/src/helpers/equal.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { isFunction, Setoid } from './function'
+
+export const equals = (a: unknown) => (b: unknown) => areEqual(a, b)
+
+const areEqual = (a: any, b: any) =>
```
Comment (ulfryk) [inline packages/core/src/helpers/equal.ts]: this function is obcolete

File: packages/core/src/helpers/empty.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+export const isEmptyObject = (value: any) =>
+  typeof value === 'object' && Object.keys(value).length === 0
+
+export const isNothing = (value: any) => value == null
+
+/* tslint:disable:no-empty */
+export const noop = () => {}
+
+export const compose = (f: Function, g: Function) =>
```
Comment (ulfryk) [inline packages/core/src/helpers/empty.ts]: duplicated

File: packages/core/src/core.ts (hunk 3 above)
Comment (rebzden) [reply packages/core/src/core.ts]: fixed

File: packages/core/src/helpers/empty.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export const isEmptyArray = (value: any) =>
+  Array.isArray(value) && value.length === 0
+
+export const isEmptyObject = (value: any) =>
+  typeof value === 'object' && Object.keys(value).length === 0
+
+export const isNothing = (value: any) => value == null
+
+/* tslint:disable:no-empty */
+export const noop = () => {}
```
Comment (ulfryk) [inline packages/core/src/helpers/empty.ts]: duplication

File: packages/core/src/helpers/function.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+/* tslint:disable:no-empty */
+export const noop = () => {};
+
+export const compose = (f: Function, g: Function) => (x: unknown) => f(g(x));
+
+export const isFunction = (f: any) =>
+  Boolean(f && f.constructor && f.call && f.apply);
+
+export interface Setoid<A> {
```
Comment (ulfryk) [inline packages/core/src/helpers/function.ts]: move it out from here to own space

File: packages/core/src/helpers/equal.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  areSimplyEqual(a, b) || !(!a || !b) || areFunctionsEqual(a, b)
+
+const areSimplyEqual = (a: any, b: any) =>
+  a === b || (isNaN(a) && isNaN(b))
+
+const areFunctionsEqual = (a: any, b: any) =>
+  isFunction(a.equals) && isFunction(b.equals) && a.equals(b)
```
Comment (ulfryk) [inline packages/core/src/helpers/equal.ts]: use `isSetoid(a)` instead

File: packages/core/src/helpers/function.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+/* tslint:disable:no-empty */
+export const noop = () => {};
+
+export const compose = (f: Function, g: Function) => (x: unknown) => f(g(x));
```
Comment (ulfryk) [inline packages/core/src/helpers/function.ts]: this one should be strongly typed

**Review by ulfryk — DISMISSED** [review]

File: packages/core/src/helpers/function.ts (hunk 11 above)
Comment (rebzden) [reply packages/core/src/helpers/function.ts]: fixed

File: packages/core/src/helpers/equal.ts (hunk 12 above)
Comment (rebzden) [reply packages/core/src/helpers/equal.ts]: fixed

File: packages/core/src/helpers/empty.ts (hunk 10 above)
Comment (rebzden) [reply packages/core/src/helpers/empty.ts]: fixed

File: packages/core/src/helpers/empty.ts (hunk 9 above)
Comment (rebzden) [reply packages/core/src/helpers/empty.ts]: fixed

File: packages/core/src/helpers/equal.ts (hunk 8 above)
Comment (rebzden) [reply packages/core/src/helpers/equal.ts]: fixed

File: packages/core/src/helpers/function.ts (hunk 13 above)
Comment (rebzden) [reply packages/core/src/helpers/function.ts]: fixed

File: packages/core/src/helpers/empty.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+export const isEmptyString = (value: any) => isNothing(value) || value === ''
+
+export const isEmptyArray = (value: any) =>
+  Array.isArray(value) && value.length === 0
+
+export const isEmptyObject = (value: any) =>
+  typeof value === 'object' && Object.keys(value).length === 0
+
+export const isNothing = (value: any) => value == null
```
Comment (ulfryk) [inline packages/core/src/helpers/empty.ts]: I'd suggest extracting `isNothing` to separate file

File: packages/core/src/helpers/equal.ts [hunk 15]
```diff
@@ -0,0 +1,9 @@
+import { isSetoid } from './setoid'
+
+export const equals = (a: unknown) => (b: unknown) => areSimplyEqual(a, b) || !(!a || !b) || areFunctionsEqual(a, b)
+
+const areSimplyEqual = (a: any, b: any) =>
+  a === b || (isNaN(a) && isNaN(b))
+
+const areFunctionsEqual = (a: any, b: any) =>
```
Comment (ulfryk) [inline packages/core/src/helpers/equal.ts]: Maybe `areSetoidsEqual`

File: packages/core/src/helpers/equal.ts [hunk 16]
```diff
@@ -0,0 +1,6 @@
+import { areSetoidsEqual } from './setoid'
```
Comment (ulfryk) [inline packages/core/src/helpers/equal.ts]: name file `equals` and add unit tests

File: packages/core/src/helpers/tests/empty.spec.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+      true,
+      false,
+      ['I\'m not empty sir!'],
+      { hello: 'there' },
+      3,
+      'Im just empty string... UPS',
+      NaN,
+    ].forEach(empty => expect(isEmpty(empty)).to.be.false)
+  })
+  given`Non empty expect value to be true`(() => {
```
Comment (ulfryk) [inline packages/core/src/helpers/tests/empty.spec.ts]: given -> when -> then
given -> expect

I'd suggest adding `expect` helper to `gherkin` helpers

so it looks like:

```typescript
let emptyValues: any[];

given `some empty values` (() => {
  emptyValues = […]
})

expect `them to pass "isEmpty" test` (() => {
  emptyValues.forEach(v => {
     xpct(isEmpty(v)).to.be.true
  })
})
```

or

```typescript
let emptyValues: any[];
let testResults: boolean[];

given `some empty values` (() => {
  emptyValues = […]
})

when `they are tested with "isEmpty"` (() => {
  testResults = emptyValues.map(isSome) // or use filter and assert count
})

then `they all pass the test` (() => {
  expect(testResults).to.deep.equal([true, true, …])
})
```

File: packages/core/src/helpers/tests/equals.spec.ts [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+  given`Not equal values expect result to be false`(() => {
+    expect(equals(true)(false)).to.equal(false)
+    expect(equals(0)(1)).to.equal(false)
+    expect(equals('yes')('no')).to.equal(false)
+    expect(equals(false)(true)).to.equal(false)
+    expect(equals([])({})).to.equal(false)
+    expect(equals({})({})).to.equal(false)
+    expect(equals(3)([3])).to.equal(false)
+    expect(equals(3)('3')).to.equal(false)
+  })
```
Comment (ulfryk) [inline packages/core/src/helpers/tests/equals.spec.ts]: Follow similar rules to ones in isEmpty

File: packages/core/src/helpers/tests/is-nothing.spec.ts [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+  given`Empty values`(() => {
+    expect(isNothing(null)).to.be.true
+    expect(isNothing(undefined)).to.be.true
+  });
+  given`Not empty values`(() => {
+    expect(isNothing(3)).to.be.false;
+    expect(isNothing(false)).to.be.false
+    expect(isNothing(true)).to.be.false
+    expect(isNothing([])).to.be.false
+    expect(isNothing({})).to.be.false
```
Comment (ulfryk) [inline packages/core/src/helpers/tests/is-nothing.spec.ts]: aa

**Review by ulfryk — APPROVED** [review]

**Pull request merged**